### ISSUE TITLE
Regression in commits from dec 24

### ISSUE THREAD

**Issue by faddat** [google/netboot#52]
Packet capture: (can filter by bootp and open in wireshark) (see packets 50 and 51)
[wireshark.zip](https://github.com/google/netboot/files/1595397/wireshark.zip)

Pixiecore log:

```
[DHCP] Offering to boot 08:00:27:4e:15:86
[DHCP] Offering to boot 08:00:27:4e:15:86
[DHCP] Offering to boot 08:00:27:4e:15:86
^[[24~[DHCP] Offering to boot 08:00:27:4e:15:86
[DHCP] Offering to boot 08:00:27:4e:15:86
[DHCP] Offering to boot 08:00:27:4e:15:86
[DHCP] Offering to boot 08:00:27:4e:15:86
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[TFTP] Sent "4c:cc:6a:ce:90:8c/7" to 192.168.1.199:1821
[TFTP] Sent "4c:cc:6a:ce:90:8c/7" to 192.168.1.199:1822
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
[TFTP] Sent "4c:cc:6a:ce:90:8c/7" to 192.168.1.199:1823
[TFTP] Sent "4c:cc:6a:ce:90:8c/7" to 192.168.1.199:1824
[DHCP] Offering to boot 4c:cc:6a:ce:90:8c
```

Machine boot log: **note: I think that this is unrelated to #51 because it happens on physical and virtual hardware alike.**
![image](https://user-images.githubusercontent.com/7142025/34463470-084b3a1a-ee2b-11e7-9870-123d365f5a5e.png)

We were able to solve this by moving back to a pixiecore built on December 17.  Next I'm going to build the bug-51 branch.  Thanks for your help, @danderson !

**Comment by danderson** [comment]
Damnit! Thanks for the report, looking. At first glance, there's no _obvious_ changes to that phase of boot, that section is just grabbing a static TFTP file. The MAC address is normal, but it's basically only for data continuity in the logs, so you can track a MAC through each stage of boot. The important part is the `/0`, which should serve an ipxe for the x86 arch...

**Comment by danderson** [comment]
Okay, looking at your pcap and screenshot, something is weird: in the pcap, Pixiecore is running on 192.168.1.192, but in your screenshot, the virtualbox firmware is trying to download the next boot stage from 192.168.1.174

I've seen this before, when there are multiple competing PXE and ProxyDHCP servers on the network. The PXE spec doesn't specify how to merge incoming responses for ProxyDHCP, so you can end up using a server IP from one response, but a filename from another... Unfortunately, I don't see 192.168.1.174 in the pcap, so I have no idea how this machine is somehow convincing your virtualbox to boot from it... 

Does this make sense to you? Is there maybe a second pixiecore/other PXE server running?

**Comment by cdata** [comment]
Hi, I'm new to both Pixiecore and PXE booting in general. As a minor anecdote that I think might be related, I was having trouble booting a small device yesterday and happened upon your conversation as it was unfolding. I tried using several of the "quick" options just to rule out any mistakes I might have made invoking Pixiecore, but the results were the same. Sure enough, rolling back to [October 17th](https://github.com/google/netboot/commit/7eee7d4832ffbf0161f51a8c03161b6b620d05d6) in pixiecore enabled my device to load everything and boot.

This may be a completely different issue, but given the title of the issue I figured I would post it in here first since i-dont-know-what-im-doing.jpg.

Attached is a pcap I took using the debug command I found in pixiecore. The device would reboot immediately upon failure to boot, but I was able to glean a reference to this error in the output: ipxe.org/3c092083

[boot.pcap.zip](https://github.com/google/netboot/files/1596117/boot.pcap.zip)

I'm not sure how else to be most helpful, but here is a basic description of my setup:

My host machine is a Macbook Pro with VirtualBox. On my host there is a VM with a bridged network adapter, and this VM runs Pixiecore. On my local network there is [a small device](https://azulletech.com/product/access-plus-fanless-windows-mini-pc-stick/) capable of booting over the network.

**Comment by danderson** [comment]
Okay, so that's two reports now that rolling back fixes things. I need to reproduce this locally somehow so I can bisect...

**Comment by danderson** [comment]
Okay! I can reproduce this in virtualbox with the architecture:

 - Virtualbox VM with no hard drive, bridged network, configured to boot from network.
 - Pixiecore running on the virtualbox host machine

I see the same symptoms, where the TFTP file inexplicably doesn't get served.

In parallel, I'm able to netboot a standard KVM machine, so it's something specific to virtualbox that's broken.

Bisecting...

**Comment by danderson** [comment]
Debugging a bit before I bisect... It looks like the virtualbox copy of ipxe never tries to speak TFTP at all, network capture only shows the DHCP exchange (which completes correctly afaict), and then the virtualbox firmware just fails without even trying to speak TFTP...

**Comment by danderson** [comment]
Verified again by comparing with a KVM VM, and I can definitely see DHCP+TFTP happening with that one.

So, a recent change made virtualbox's strange version of ipxe unhappy. Digging...

**Comment by danderson** [comment]
Found one solution: you need to install the virtualbox extension pack at https://www.virtualbox.org/wiki/Downloads . According to various internet sources, PXE boot from internal iPXE when using an emulated intel NIC is known-broken, and the "fix" is to install the extension pack, which provides the standard intel PXE ROM. This ROM can boot correctly, no problem.

I'm not sure I can fix virtualbox's builtin borked ipxe :( Their iPXE claims that it supports TFTP, but when I try and make it speak TFTP, it just fails without ever connecting to anything, with an error that makes no sense. So, that's not encouraging... I'll keep digging, but it's possible the "solution" will just be "don't use virtualbox, or if you have to use virtualbox, install the extension pack"

@cdata I suspect your own problem is something different, based on the fact that your setup is reversed (running pixiecore on virtualbox, booting a physical system). I'll investigate your report as soon as I close this one with a documentation fix.

**Comment by faddat** [comment]
@danderson for us it was not only VirtualBox that was broken. We had the same issue (mac address of client showing up as a filename) when booting using bare metal, too.

**Comment by danderson** [comment]
The MAC address in the filename is normal, we tell the firmware to boot /<mac-addr>/<arch-number>, so that the TFTP server knows which machine is requesting a boot file. You should see that in all boot attempts, and it shouldn't be a problem. Are you seeing problems with booting bare metal as well (no VirtualBox anywhere in the setup) ?

**Comment by danderson** [comment]
Okay, I think I've found what makes the embedded virtualbox iPXE unhappy. For some reason, it doesn't like the standard PXE format of specifying next-server+filename as a way of saying "you should boot with `tftp://<next-server>/<filename>`. Instead, it tries to interpret the filename `<mac>/<arch>` as a boot URL, and fails because it's not a URL.

I verified manually that if I give it a "filename" of `tftp://<next-server>/<mac>/<arch>`, virtualbox's ipxe is capable of chainloading to Pixiecore's ipxe, and from there we can successfully boot.

So, I think the change we need is:

 - Always chainload to Pixiecore's iPXE, so we have a known set of features.
 - When we want to chainload, detect if the PXE firmware is iPXE or non-iPXE. If iPXE, set the filename to `tftp://<next-server>/<mac>/<arch>`. Otherwise, set the filename to `<mac>/<arch>`.

This will require some refactoring of the boot logic, so it'll take a bit of time to make the change. I'll report back once it's done.

**Comment by danderson** [comment]
Okay, new problem... I implemented this, but now after chainloading to Pixiecore's iPXE, it cannot configure the network any more... iPXE's website says this can be a problem when you chainload undionly.kpxe (which is what we chainload), due to buggy firmwares :(

**Comment by danderson** [comment]
So, recapping, the failure you see here is actually a result of both of my attempts to fix #51. The obvious way to fix the problem triggered a bunch of bugs in my understanding of ipxe, and ipxe itself. I've updated #51 with the plan to fix Virtualbox booting.

Unclear if @cdata's issue is the same. The error he's getting suggests that it is, and if so, it's possible my virtualbox fix will also fix that. I'm filing a new bug to investigate that.
